### DISCUSSION TITLE
Progress bar for graph creation

### DISCUSSION THREAD

**Discussion by acjones27** [networkx/networkx#6328] [Ideas]
Hi,

I'm building a pretty big graph (like 1 billion edges) but I never know how long it's going to take (hours, days, or longer.. ) to create the graph. Does anyone know of a way to see the progress? Even if there were a way to know how far along it was when I killed the process or it died of natural causes e.g. I waited 2 days and only 1% of the edges had been added.

Any suggestions? Could this feature be added?

**Comment by SultanOrazbayev** [comment]
How are you creating the graph? Using `tqdm` as a wrapper is often a workable solution.

**Comment by dschult** [comment]
Every progress bar has to have a way to predict how long the process will take. In this case, even if you track how many edges have been added, it will depend on the type of graph being created.

Probably the best way to predict is to build a bunch of smaller graphs using the same creation algorithm, while increasing the size of the graph consistently.  For example, if you double the number of nodes and take the ratio of the times, you will get 2^p when the algorithm is O(n^p) for n the number of nodes. If the dependence is faster than n^p for any p then you can explore O(a^n) behavior by increasing the number of nodes by 1 and computing the ratio of times to get a. The estimates of a and p will only be valid if they stay approximately constant in the limit of large n. 

The timing dependence on n is also only a valid prediction if you aren't running into hardware bottlenecks. Swapping from memory to hard drive space can really slow down any memory intensive computations. If you are taking days to create a graph with a billion edges, then I would suspect memory issues are a potential bottleneck.

You might also consider that graphs of this size may not be what you really need for the question you are asking. But that's a whole other discussion... :}

**Comment by acjones27** [comment]
@SultanOrazbayev I don't think `tdqm` is an option since I'm creating it with the following code (so nothing for `tdqm` to iterate over):

```python
G = cugraph.DiGraph()
G.from_dask_cudf_edgelist(e_list)
```

@dschult Thanks for the advice! Yes, to be honest I've switched to using NVIDIA `cugraph` now because I think it's just too big for `networkx` on CPU. But also `cugraph` doesn't have a progress bar so I'm flying blind.. For more context, I was able to load a 500M-edge graph in `cugraph` in under a minute on a single A100 GPU with 40GB of memory. But once you scale up to 1.7Bn edges, you need multiple GPUs and then there's the issue of latency with swapping the memory as you mentioned (plus the added cost, which is scary when you don't know how long it will be running for... )

As for whether a ginormous graph is really the answer, I'm trying to calculate pagerank so I couldn't think of another way other than building the whole graph. And I can't really build a smaller graph as each node's pagerank influences the other nodes that it connects to.

**Reply by MridulS** [reply]
If you are in a 1.7bn graph, `cugraph` is probably the right answer. Now how to get this data on multiple GPUs is probably better answered by cuGraph folks. Maybe @eriknw has some pointers?

**Reply by eriknw** [reply]
Hey there!

For doing PageRank on a very large graph, `cugraph` is a good choice (if you have access to GPUs). Note that cuGraph's uses the parallel power method for PageRank, so its values may be different from what you expect, but the _ranking_ of vertices ought to be correct.

Here's a notebook showing a simple example with PageRank:
https://github.com/rapidsai/cugraph/blob/branch-23.02/notebooks/algorithms/link_analysis/Pagerank.ipynb

and here's a notebook running PageRank on multiple GPUs:
https://github.com/rapidsai/cugraph/blob/branch-23.02/notebooks/demo/mg_pagerank.ipynb

**Comment by dschult** [comment]
To the extent that "flying blind" is a problem, the answer comes from using smaller graphs to get an estimate of how the time changes as you change the size. Moving from 0.5B to 1.4B edges won't allow you to predict the time needed as well as 0.01B edges to 0.02B edges 0.04B edges to 0.08B edges. The hope is that you can extrapolate that time dependence to the larger numbers of edges. It can often at least give you a ballpark number (15 min or 15 hours or 15 days). :}

If you are going for computing pagerank on the graph, cugraph is a good choice.  But that shouldn't mean you can't split the network into sections. Google certainly has to do that to compute pagerank on the network of webpages. But my understanding is that they gave up on eigenvector solvers in favor of Monte Carlo simulations of randomly moving particles with the eigenvector values given by portion of particles ending up at each node. Cool stuff.... and I think you are still small enough that it isn't needed. :}

**Comment by dschult** [comment]
GraphBLAS would be another good solution.  And they have alternate backends that work well with the development version of NetworkX.  @eriknw can maybe give some hints on that as well.  :}

**Reply by acjones27** [reply]
Ok thanks, I'll check that out! :)

**Reply by eriknw** [reply]
I think GraphBLAS is a great choice for very large PageRank on CPU if you have a beefy machine with lots of memory. We implement PageRank (and many other NetworkX functions) in [`graphblas-algorithms`](https://github.com/python-graphblas/graphblas-algorithms/). We recently gave a talk ([here](https://youtu.be/pd-Sbm8lHTc)) about NetworkX 3.0 dispatching to GraphBLAS (via `graphblas-algorithms` and [`python-graphblas`](https://github.com/python-graphblas/python-graphblas)), and here are some results for how much faster GraphBLAS was for me:
![graphblas_vs_networkx](https://user-images.githubusercontent.com/2058401/210926345-0fe89931-22da-4bd3-af99-a9a53bf9e529.PNG)

For the larger datasets, GraphBLAS was typically >20x faster than `scipy.sparse` implementation of PageRank.

We're [working on](https://github.com/metagraph-dev/dask-grblas/pull/21) a distributed version of `python-graphblas` called `dask-grblas` / `dask-graphblas` (it's being renamed to the latter).  We haven't updated or released it in a long time, so installing it and running it may not be super-easy. But, it should work.  @jim22k shared code here: https://github.com/dask/dask/discussions/9721#discussioncomment-4343847

If you have any questions for us about GraphBLAS, feel free to [join our weekly call](https://github.com/python-graphblas/python-graphblas/issues/247) or create [issues](https://github.com/python-graphblas/python-graphblas/issues) or continue in this thread.
